core/src/main/java/io/grpc/SynchronizationContext.java

```diff
@@ -0,0 +1,203 @@
+   * {@link #executeLater} immediately followed by {@link #drain}.
+   */
+  @Override
+  public final void execute(Runnable task) {
+    executeLater(task);
+    drain();
+  }
+
+  /**
+   * Schedules a task to be run via {@link #execute} after a delay.
+   *
+   * @param task the task being scheduled
+   * @param delay the delay
+   * @param unit the time unit for the delay
+   * @param timerService the {@code ScheduledExecutorService} that provides delayed execution
+   *
+   * @return an object for checking the status and/or cancel the scheduled task
+   */
+  public final ScheduledHandle schedule(
+      final Runnable task, long delay, TimeUnit unit, ScheduledExecutorService timerService) {
+    final ManagedRunnable runnable = new ManagedRunnable(task);
+    ScheduledFuture<?> future = timerService.schedule(new Runnable() {
+        @Override
+        public void run() {
+          execute(runnable);
+        }
+
+        @Override
+        public String toString() {
+          return task.toString() + "(scheduled in " + SynchronizationContext.this + ")";
+        }
+      }, delay, unit);
+    return new ScheduledHandle(runnable, future);
+  }
+
+  private static class ManagedRunnable implements Runnable {
+    final Runnable task;
+    boolean isCancelled;
+    boolean hasStarted;
+
+    ManagedRunnable(Runnable task) {
+      this.task = checkNotNull(task, "task");
+    }
+
+    @Override
+    public void run() {
+      // The task may have been cancelled after timerService calls SynchronizationContext.execute()
+      // but before the runnable is actually run.  We must guarantee that the task will not be run
+      // in this case.
+      if (!isCancelled) {
+        hasStarted = true;
+        task.run();
+      }
+    }
+  }
+
+  /**
+   * Allows the user to check the status and/or cancel a task scheduled by {@link #schedule}.
+   *
+   * <p>This class is NOT thread-safe.  All methods must be run from the same {@link
+   * SynchronizationContext} as which the task was scheduled in.
+   */
+  public static final class ScheduledHandle {
+    private final ManagedRunnable runnable;
+    private final ScheduledFuture<?> future;
+
+    private ScheduledHandle(ManagedRunnable runnable, ScheduledFuture<?> future) {
+      this.runnable = checkNotNull(runnable, "runnable");
+      this.future = checkNotNull(future, "future");
+    }
+
+    /**
+     * Cancel the task if it's still {@link #isPending pending}.
+     */
+    public void cancel() {
+      runnable.isCancelled = true;
+      future.cancel(false);
+    }
+
+    /**
+     * Returns {@code true} if the task will eventually run, meaning that it has neither started
+     * running nor been cancelled.
+     */
+    public boolean isPending() {
+      return !(runnable.hasStarted || runnable.isCancelled);
+    }
+  }
+
+  /**
+   * Handle a throwable from a task.
+   *
+   * <p>The default implementation logs a warning.
+   */
+  protected void handleUncaughtThrowable(Throwable t) {
```

Comment:
Nit: the class could be made final and be passed a Thread.UncaughtExceptionHandler (with a note that the thread will not die after executing the handler, which is different from its documentation).

Reply:
Done.